### ISSUE TITLE
Demonstrate how temporalization is able to integrate both normal and depletion updates

### ISSUE THREAD

**Issue by pedrogit** [edwardsmarc/CASFRI#315]

**Comment by pedrogit** [comment]
According to SC a "depletion inventory update" is different from a "normal inventory update" in that it contains only disturbance information. The question is how to link this information to the previous stand containing species information.

Our guess is that depleted polygons are implicitely linked to older polygons spatially. If one require older stand information about a polygon coming from a depleted inventory, he has to query the historical database to get the older polygon having the same spatial footprint.

**Comment by kdarndt** [comment]
We are interested in how this is approached as well.  If a "depletion inventory" is simply a subset of a regular inventory with stand attributes updated for depletion, it seems like this could be incorporated into CASFRI fairly easily.  
If, however, a "depletion inventory update" is simply information on a disturbance where CASFRI must then infer the resulting stand attributes, that seems much more complex.

I think we would want someone querying at a point in time to get the most complete picture of the stand (including stand attributes updated for depletion, if appropriate for that time/place).

**Comment by pedrogit** [comment]
Just thought about the fact that many inventories have records carrying ONLY disturbances info (no LYR or NFL info). In what are those records different from records from a so called "depletion update" if those "depletion updates" are composed only of disturbance info? That means we already have the problem of how to infer the type of forest in place before a disturbance in current inventories.

**Comment by MelinaHoule** [comment]
CASFRI v4 used to deal with up to 3 disturbances only. By implementing the temporalization with the inclusion of depletion updates, we wanted to expand unlimited historical information. In many jurisdictions, there are produced annually to “cut-in” polygons disturbed by harvesting or wildfire. When an update of disturbance occurs in a stand, CASFRI should store the vegetation class before the disturbance happens. **In its current state, CASFRI does not keep this information.** The temporalization would allow to keep information on vegetation through time and allow to cumulate disturbance history.  For example, a big polygon A of 100% black spruce interpreted in 2014. We get the depletion update of cut over (polygon B) in 2018 and compare the geometries. Polygon B intersect polygon A (donut shape). Valid time from polygon A would be 2014-2018, 100% black spruce. And starting from 2018 and up, it would be 2 geometries, the donut (old A) 100% black spruce, and the donut hole (depletion update, polygon B) with  no LYR but cut block disturbance and the user  will get the history by looking at VALID TIME START and VALID TIME END.  If the user wants to know the state of the forest at a precise time, we assume they are not interested to know what was there in the past, so the information is not gonna go along. 

I always see that depletion updates would be treated as new inventories since there is no chances the geometry would fit and I don't see how a user could work with a polygon 100% black spruce with valid time 2014-2020 that was cut in 2018, so the valid time of LYR is not that valid because you need to look into disturbance if there are something that came up between. It makes the information really hard to parse for analysis compare to multiple polygons with valid time well set.

**Comment by pedrogit** [comment]
What Melina described is what is intended to do. Implementing "completion" would mean that the donut hole in Melina's example would "inherit" polygon's A LYR info. Thus both polygons, in the final historical database would have LYR info.

Without disturbance info a forest inventory is basically a "state" database, in which the forested (or non forested) state of the forest is valid from one period of time (VALID_YEAR_BEGIN) to another period of time (VALID_YEAR_END). In the current CASFRI schema, non-forested and forested info are modeled in separate tables even if they both represent the state of a stand. The new schema merge those two type of info into a single table called subcomponent. The point here is that both type of info represent the current state of the stand.

Disturbance info is different. It does not represent "state". It represent "change". It should be modeled as any other "state\change" database.

The problem we have here is that disturbance is modeled in two different ways: 

1) In traditional inventories, "change" (disturbance) is modeled as additional attributes to "states" (forest info).

2) In depletion update inventories, only "change" is modeled. (Note that in traditional inventories, disturbances are also very often modeled alone without forest info as we can see in CASFRI where many record have disturbance info only.

So now how should "change" be modeled in a "state" database?

Think about any edition software (PowerPoint, PhotoShop, Word, etc..) were you are modifying the state of the data you are editing. What you see is the actual "state" of the data. The software keeps all the history of your "changes" so that you can always revert to previous "states" by undoing the "change". Clearly two very different kinds of information has to be modeled and kept to enable this kind of functionality: current "state" and history of "changes". In an alternative model, you could also rather only keep all the history of "states" and get rid of the history of "changes" but this requires more memory because each historical "state" contains much more information than each "change". That's why historically, the first model has been adopted even if going back to a precise moment in time requires more computing (you have to undo each "change" one by one).

The advantage of the second model ("history of states") over the first model ("current state and history of changes") is that it does not require any piece of software (or algorithm) to "compute" historical states. They are all contained in the database in a determined order. It also does not require to keep the history of changes. You can theoretically get rid of them. That's CASFRI: it does not provide any piece of software to compute "historical states" from the "history of change". Instead it provides all historical "states". Querying an historical state is just a matter of querying the database at a specific time.

Hence every "change" introduce a new "state" in the "history of states". That means each disturbance, instead of being stored along the former forest info (isn't it odd to store that a forest state is composed of something when it has actually been completely cut?), should introduce a new forest record ("state") applying the disturbance ("change") from the previous record in history. The result is the actual "state" of the forest at any time. Theoretically we could then get rid of disturbance info since it is always reflected in the historical states, but we like to keep it along the actual state because we like to know what is the origin of the state.

The difference with the actual model is that instead of having many (generally up to three) "future" disturbances stored along a forest state, there would only be one disturbance per state. The three disturbances associated with a stand would still exists along the new states generated by the disturbances. We could now see those compagnon disturbances more as "origin of the state" info (clear cut, fire, windfall, etc... we could introduce "normal succession" for stands not resulting from disturbances).

In Melina's example there would be three resulting polygons: 

1- one LYR polygon representing the first state (LYR: 100% black spruce, origin: succession, valid from 2014 to 2017), 
2- one new LYR polygon (the donut hole) representing the cut (no LYR info, origin: cut, valid from 2014 to say 2020)
3- one new polygon (the donut) representing the stand part not affected by the disturbance (same info as polygon 1)

The interest of this model is that disturbance info in integrated into a single, simple table, representing the history of forest states with their forest or non forest characteristics and the cause of their origin. Within the new schema that would be the subcomponent table (with the addition of "origin of state" info). No more disturbance table needed.

**Comment by kdarndt** [comment]
I agree that CASFRI should be a "history of states".  We don't wan to build an algorithm to compute "state" (forest attributes) as a result of a "change" (ie: disturbance).  Let's leave that up to the jurisdictions.  
I was under the assumption that a "depletion update" contains only "state" as a result of "change" (they have already computed forest attributes as a result of disturbance) -- need clarification on this.

However, there is a distinction between:
 1) "no data" and 
 2) "there is no more black spruce because it has been cut down". 

For (1) it might be advantageous to carry forward information from an older polygon.  For example, if a depletion update indicates that there is 0% black spruce in the donut "hole", but leaves the NFL information "blank" (no data) -- can we or should we carry forward information on NFL?  
I think it depends on whether the jurisdiction supplies information on only those attributes affected by a disturbance (ie: a fire where the non-forested layer (NFL) was untouched?).

For (2) the values for the Forested layer (LYR) should not be blank -- but should indicate that there is no forest.  I'm not sure if this is typically done in inventories, but it provides information on the state whereas "no data" does not.






 In the latter, the "donut hole" (using Melina's example) should have the LYR attributes populated with information indicating 
However, there is still a question of missing attributes and how to deal with them:
 - if a disturbance causes the donut "hole" to no longer be 100% black spruce, the CASFRI attributes for the

**Comment by edwardsmarc** [comment]
Assuming all depletion update polygons will eventually be replaced by a re-inventory, is the following scenario accurate?

From Melina's example, lets say in 2025 a complete new inventory is released and the polygon that was cut (the donut hole) is now labelled as 100% black spruce again (i.e. it grew back after the fire). Queries on the LYR for that polygon would return the following:

* 2014 to 2017 - 100% black spruce
* 2017 to 2025 - unknown
* 2025+ - 100% black spruce

**Comment by edwardsmarc** [comment]
Here is an old document with some info on depletion updates that may be useful:


[Forest Vegetation Inventory Updates in Canada.docx](https://github.com/edwardsmarc/CASFRI/files/4629724/Forest.Vegetation.Inventory.Updates.in.Canada.docx)

**Comment by sgautam-nrcan** [comment]
I think if the inventory completely updated based on the new interpretation (as discussed in the above example 2025) it will update. However, the question still there how to handle if the forest was cut in 2017.  In this case, the above-discussed methods "history of states" sounds good. If the history of states provides the location and intensity of disturbance (by jurisdiction) then it could be used to adjust into older values. But if the location and intensity of disturbance are not clear (e.g., percentage of removal for a polygon) then it could be difficult.

**Comment by MelinaHoule** [comment]
> Assuming all depletion update polygons will eventually be replaced by a re-inventory, is the following scenario accurate?
> 
> From Melina's example, lets say in 2025 a complete new inventory is released and the polygon that was cut (the donut hole) is now labelled as 100% black spruce again (i.e. it grew back after the fire). Queries on the LYR for that polygon would return the following:
> 
> * 2014 to 2017 - 100% black spruce
> * 2017 to 2025 - unknown
> * 2025+ - 100% black spruce

**Comment by MelinaHoule** [comment]
To add a comment on Marc's answer, queries on LYR in 2025 for a 2014 polygon being disturbed in 2018 would return the following :

2014 to 2017 - 100% black spruce
**2017 to 2025 - Potentially Productive**
2025+ - 100% black spruce

**Comment by edwardsmarc** [comment]
Conclusions from our meeting today:

_Item 1. The meaning & expectations of the deliverable “a CASFRI query script that combines the best (most complete attribution) and most relevant (inventory date closest to user-specified date of interest) available FRI data into a topologically clean dataset with no overlaps for the user-specified area of interest.”_ 

**We agreed that “most complete attribution” simply means choosing the most attribute-rich polygon (ie: among overlapping polygons within the same year) within the historical database.  No attributes from different polygons will be carried “forward” or “backward” in time to fill in gaps in other polygons.**

 

_Item 2. Incorporating “depletion” inventories into CASFRI.  Depending on the nature of depletion inventories, this topic may be related to topic (1)._

**Depletion updates and disturbance information contained in regular inventories are desirable in CASFRI.  We want a user, when querying the historical database, to know when a disturbance has affected a particular polygon and be able to also query the stand information for that polygon as it existed prior to the disturbance.  There are several technical ways to implement this which were not fully discussed in the call as that was out of scope of this meeting.  Disturbance information and regular inventory information will remain separate in the “raw” (non-historical) CASFRI tables.**

**Comment by edwardsmarc** [comment]
Notes from a Slack conversation between @edwardsmarc , @pedrogit and @kdarndt:

### Distinction between PHOTO_YEAR and DIST_YEAR
- PHOTO_YEAR is the year the photo was taken for an inventory, or the year of publication for a disturbance-only post-inventory update (see #473).
- Since the historical database is built using PHOTO_YEAR and not DIST_YEAR, we are really querying for the **state of the inventory**, not the **state of the forest**.
- e.g. we have an inventory with photo year 2000, then a second inventory for the region with a photo year of 2010. In 2001 there was a cutblock which is listed as a disturbance in the 2010 dataset. **We all agree that CASFRI v5 will not edit the LYR/NFL records to account for this disturbance. The only LYR/NFL information available are the records from 2000 and 2010. It's up to the historical database to return the most relevant LYR/NFL info based on the query, but we won't manually create a LYR record in 2001 to reflect the altered state of the forest.**
  - This is what we refer to as *state of the inventory*. If we wanted the database to return the *state of the forest* we would have to create a new LYR/NFL record for every DST record in the database. This would be like manually interpreting every disturbance and creating a new LYR record reflecting the change to the forest. This is outside the scope of the current project where we have agreed we will only report the information that is available.

### Distinction between depletion update and disturbance-only update
- Depletion update: this is a set of disturbance polygons (e.g. cutblocks) released post-inventory where the LYR and NFL attributes have been updated to reflect the disturbance.
  - An update cycle for a depletion update would be:
    1. First full inventory dataset with PHOTO_YEAR = x
    2. Depletion update released with PHOTO_YEAR = y
    3. Second full inventory released, absorbing the disturbances from the depletion update, with PHOTO_YEAR = z.
  - We can just treat depletion updates as partial inventories since they have valid PHOTO_YEAR.
- Disturbance-only updates: this is a set of disturbance polygons (e.g. cutblocks) released post-inventory, but there is no update for the LYR and NFL attributes. The only information if the disturbance type, and disturbance year. 
  - An update cycle for a disturbance update would be:
    1. First full inventory dataset with PHOTO_YEAR = x
    2. Disturbance update released with PHOTO_YEAR = UNKNOWN
    3. Second full inventory released, absorbing the disturbances from the disturbance update, with PHOTO_YEAR = z.
  - We cannot treat disturbance updates as inventories because they do not have a PHOTO_YEAR. They will be added to the database but are a low priority.

### Wrap up
- It is not possible to create a historical database that can query the *state of the forest* using **only** the source inventory records.
- In order to query the *state of the forest* CASFRI would have to compute an updated LYR record for every reported disturbance event.
- In the absence of this, our historical database will return the most relevant information based on what is available from the source inventories.

### Useful comments from @pedrogit:
"Let's explain this another way: we have two very different timestamps in the inventories: 1) PHOTO_YEAR is a good indicator of the validity of LYR and NFL info. Is is useful to determine the state of the forest at this year in the historical database. 2) There is is also DIST_YEAR_1, DIST_YEAR_2 and DIST_YEAR_3 which is a historical list of past disturbances. These years are NOT used as historical markers in the historical database and hence are not useful (yet) to determine the state of the forest. i.e. if you search for data in 2000 you might miss a disturbance that was recorded in the inventory only 5 or 10 years later.
So I would say that with the current historical database you get a partial view of the "state" of the forest (LYR and NFL info is good but DST info as to be more precisely searched for). If we would integrate (or apply) the disturbances ("the changes") in the current or the new schema by creating new LYR rows, then we could get a complete state of the forest when querying the historical database with a given year."

"I agree that "the state of the forest" is a way more interesting dataset then "the state of the inventory" but we are not there yet and what we agreed on in our special meeting does not allow yet to produce "the state of the forest". All those things (the issues with the current schema, the way to deal with DST and depletion updates, the limits of the historical database) are becoming clearer and clearer to me. I think we got stock with a schema too much influenced by existing provincial schemas (mostly western ones). I think now we have a way better idea about how to model forest inventories for querying purpose and this is where CASFRI should go in the future. Forest inventories were primarily design for forest management. I think CASFRI should be more targeted at being a "state of the forest" ecological dataset for research."
